rts/motoko-rts/src/stream.rs

```diff
@@ -0,0 +1,184 @@
+        unsafe {
+            let next_ptr64 = (*self).ptr64 + n.as_u32() as u64;
+            stable64_write_moc((*self).ptr64, ptr as u64, n.as_u32() as u64);
+            (*self).ptr64 = next_ptr64
+        }
+    }
+
+    #[cfg(feature = "ic")]
+    /// Sets up the bottleneck routine to output towards a range of stable memory
+    /// Note: assumes that the entire byte range is writable
+    #[export_name = "stream_stable_dest"]
+    pub fn setup_stable_dest(self: *mut Self, start: u64, limit: u64) {
+        unsafe {
+            (*self).ptr64 = start;
+            (*self).limit64 = limit;
+            (*self).outputter = Self::send_to_stable;
+        }
+    }
+
+    /// Ingest a number of bytes into the stream.
+    #[export_name = "stream_write"]
+    pub fn cache_bytes(self: *mut Self, ptr: *const u8, n: Bytes<u32>) {
+        unsafe {
+            if (*self).limit64 != 0 && n > STREAM_CHUNK_SIZE
+                || (*self).filled + n > (*self).header.len
+            {
+                self.flush();
+                ((*self).outputter)(self, ptr, n);
+            } else {
+                let dest = self
+                    .as_blob_mut()
+                    .payload_addr()
+                    .add((*self).filled.as_usize());
+                (*self).filled += n;
+                assert!((*self).filled <= (*self).header.len);
+                memcpy_bytes(dest as usize, ptr as usize, n);
+            }
+        }
+    }
+
+    /// Ingest a single byte into the stream.
+    #[export_name = "stream_write_byte"]
+    pub fn cache_byte(self: *mut Self, byte: u8) {
+        unsafe {
+            if (*self).filled >= (*self).header.len {
+                self.flush()
+            }
+            self.as_blob_mut().set((*self).filled.as_u32(), byte);
+            (*self).filled += Bytes(1)
+        }
+    }
+
+    /// Return a pointer to a reserved area of the cache and advance the
+    /// fill indicator beyond it.
+    #[export_name = "stream_reserve"]
+    pub fn reserve(self: *mut Self, bytes: Bytes<u32>) -> *mut u8 {
+        unsafe {
+            if (*self).filled + bytes > (*self).header.len {
+                self.flush()
+            }
+            let ptr = self
+                .as_blob_mut()
+                .payload_addr()
+                .add((*self).filled.as_usize());
+            (*self).filled += bytes;
+            ptr
+        }
+    }
+
+    /// Split the stream object into two `Blob`s, a front-runner (small) one
+    /// and a latter one that comprises the current amount of the cached bytes.
+    /// Lengths are adjusted correspondingly.
+    #[export_name = "stream_split"]
+    pub unsafe fn split(self: *mut Self) -> Value {
+        (*self).header.len = INITIAL_STREAM_FILLED - size_of::<Blob>().to_bytes();
+        (*self).filled -= INITIAL_STREAM_FILLED;
+        let blob = (self.cache_addr() as *mut Blob).sub(1);
+        (*blob).header.tag = TAG_BLOB;
+        debug_assert_eq!(blob.len(), (*self).filled);
```

Comment:
Is this actually correct. Isn't the extent of a blob the len rounded up to the nearest word? If blob.len is less than that (because we only filled a small portion of the cache, then there will be some gap in the heap, I think.

Reply:
I am pretty sure that `len` can be 1. Otherwise how would you have `"a" : Text`?

Reply:
Right, but I don't think the blob's physical size should be more than 3-bytes longer than len, right? Perhaps it doesn't matter at the moment, but I think it might create gaps in the heap that don't contain objects.

Reply:
The code is correct. Before:
```
+---- *mut Stream
|
V
┌────────────┬─────┬───────┬─────────┬───────────┬────────┬──────────┐
│ tag (blob) │ len │ ptr64 │ limit64 │ outputter │ filled │ cache... │
└────────────┴─────┴───┴───┴────┴────┴───────────┴────────┴──────────┘
```
After
```
┌────────────┬─────┬───────┬─────────┬────────────┬────────┬─────────┐
│ tag (blob) │ 16  │ ptr64 │ limit64 │ tag (blob) │  len   │ viscera │
└────────────┴─────┴───┴───┴────┴────┴────────────┴────────┴─────────┘
                                     ^
                                     |
                                     +------- result blob
```

Reply:
Aaaah, I see what you mean. You are concerned about the empty space _after_ `viscera`, which could indeed be shorter than the original thing. It would be easy to fill that with another _valid_ heap object too.
_But_ after `Stabilisation` the heap gets erased anyway...

Reply:
yes, I guess what I was trying to say is that the physical size of the second blob will be BLOB+len'+3/4 (as seen by the GC) and thus there may be some junk in the heap (your ###). What I don't know is whether that is problematic or not - do we have a one word GC object that we could use to fill the space? I imagine if we implement a mark sweep algorithm, the junk might be problematic.

Reply:
Yep, I guess we can reuse this: https://github.com/dfinity/motoko/blob/5319cd5dd87bc9a43a27a048b44cd5eea7336fd7/rts/motoko-rts/src/types.rs#L476

Reply:
perfect.

Reply:
Okay, bfee93a62 should take care of this. PHAL.